File: app/src/main/java/com/asfoundation/wallet/ui/iab/AppcoinsRewardsBuyPresenter.java
package com.asfoundation.wallet.ui.iab;

import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import com.appcoins.wallet.appcoins.rewards.Transaction;
import com.appcoins.wallet.billing.repository.entity.TransactionData;
import com.asf.wallet.R;
import com.asfoundation.wallet.billing.analytics.BillingAnalytics;
import com.asfoundation.wallet.entity.TransactionBuilder;
import com.asfoundation.wallet.util.CurrencyFormatUtils;
import com.asfoundation.wallet.util.TransferParser;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import java.math.BigDecimal;
import java.util.concurrent.TimeUnit;

import static com.asfoundation.wallet.analytics.FacebookEventLogger.EVENT_REVENUE_CURRENCY;
import static com.asfoundation.wallet.ui.iab.RewardsManager.RewardPayment.Status.FORBIDDEN;

public class AppcoinsRewardsBuyPresenter {
  private final AppcoinsRewardsBuyView view;
  private final RewardsManager rewardsManager;
  private final Scheduler viewScheduler;
  private final CompositeDisposable disposables;
  private final BigDecimal amount;
  private final String uri;
  private final String packageName;
  private final TransferParser transferParser;
  private final boolean isBds;
  private final BillingAnalytics analytics;
  private final TransactionBuilder transactionBuilder;
  private final CurrencyFormatUtils formatter;
  private final int gamificationLevel;
  private final AppcoinsRewardsBuyInteract appcoinsRewardsBuyInteract;

  AppcoinsRewardsBuyPresenter(AppcoinsRewardsBuyView view, RewardsManager rewardsManager,
      Scheduler viewScheduler, CompositeDisposable disposables, BigDecimal amount, String uri,
      String packageName, TransferParser transferParser, boolean isBds, BillingAnalytics analytics,
      TransactionBuilder transactionBuilder, CurrencyFormatUtils formatter, int gamificationLevel,
      AppcoinsRewardsBuyInteract appcoinsRewardsBuyInteract) {
    this.view = view;
    this.rewardsManager = rewardsManager;
    this.viewScheduler = viewScheduler;
    this.disposables = disposables;
    this.amount = amount;
    this.uri = uri;
    this.packageName = packageName;
    this.transferParser = transferParser;
    this.isBds = isBds;
    this.analytics = analytics;
    this.transactionBuilder = transactionBuilder;
    this.formatter = formatter;
    this.gamificationLevel = gamificationLevel;
    this.appcoinsRewardsBuyInteract = appcoinsRewardsBuyInteract;
  }

  public void present() {
    view.lockRotation();
    handleBuyClick();
    handleOkErrorClick();
    handleSupportClicks();
  }

  private void handleOkErrorClick() {
    disposables.add(view.getOkErrorClick()
        .subscribe(__ -> view.errorClose()));
  }

  private void handleBuyClick() {
    disposables.add(transferParser.parse(uri)
        .flatMapCompletable(transaction -> rewardsManager.pay(transaction.getSkuId(), amount,
            transaction.toAddress(), packageName, getOrigin(isBds, transaction),
            transaction.getType(), transaction.getPayload(), transaction.getCallbackUrl(),
            transaction.getOrderReference(), transaction.getReferrerUrl())
            .andThen(rewardsManager.getPaymentStatus(packageName, transaction.getSkuId(),
                transaction.amount()))
            .observeOn(viewScheduler)
            .flatMapCompletable(
                paymentStatus -> handlePaymentStatus(paymentStatus, transaction.getSkuId(),
                    transaction.amount())))
        .doOnSubscribe(disposable -> view.showLoading())
        .subscribe());
  }

  @Nullable private String getOrigin(boolean isBds, TransactionBuilder transaction) {
    if (transaction.getOrigin() == null) {
      return isBds ? "BDS" : null;
    } else {
      return transaction.getOrigin();
    }
  }

  private Completable handlePaymentStatus(RewardsManager.RewardPayment transaction, String sku,
      BigDecimal amount) {
    sendPaymentErrorEvent(transaction);
    switch (transaction.getStatus()) {
      case PROCESSING:
        return Completable.fromAction(view::showLoading);
      case COMPLETED:
        if (isBds && transactionBuilder.getType()
            .equalsIgnoreCase(TransactionData.TransactionType.INAPP.name())) {
          return rewardsManager.getPaymentCompleted(packageName, sku)
              .flatMapCompletable(purchase -> Completable.fromAction(view::showTransactionCompleted)
                  .subscribeOn(viewScheduler)
                  .andThen(Completable.timer(view.getAnimationDuration(), TimeUnit.MILLISECONDS))
                  .andThen(
                      Completable.fromAction(appcoinsRewardsBuyInteract::removeAsyncLocalPayment))
                  .andThen(Completable.fromAction(
                      () -> view.finish(purchase, transaction.getOrderReference()))))
              .observeOn(viewScheduler)
              .onErrorResumeNext(throwable -> Completable.fromAction(() -> {
                throwable.printStackTrace();
                view.showGenericError();
                view.hideLoading();
              }));
        }
        return rewardsManager.getTransaction(packageName, sku, amount)
            .firstOrError()
            .map(Transaction::getTxId)
            .doOnSuccess(view::finish)
            .ignoreElement();
      case ERROR:
        return Completable.fromAction(view::showGenericError);
      case FORBIDDEN:
        return Completable.fromAction(() -> view.showError(mapError(FORBIDDEN)));
      case NO_NETWORK:
        return Completable.fromAction(() -> {
          view.showNoNetworkError();
          view.hideLoading();
        });
    }
    return Completable.error(new UnsupportedOperationException(
        "Transaction status " + transaction.getStatus() + " not supported"));
  }

  public void stop() {
    disposables.clear();
  }

  void sendPaymentEvent() {
    analytics.sendPaymentEvent(packageName, transactionBuilder.getSkuId(),
        transactionBuilder.amount()
            .toString(), BillingAnalytics.PAYMENT_METHOD_REWARDS, transactionBuilder.getType());
  }

  void sendRevenueEvent() {
    analytics.sendRevenueEvent(formatter.scaleFiat(appcoinsRewardsBuyInteract.convertToFiat(
        transactionBuilder.amount()
            .doubleValue(), EVENT_REVENUE_CURRENCY)
        .blockingGet()
        .getAmount())
        .toString());
  }

  void sendPaymentSuccessEvent() {
    analytics.sendPaymentSuccessEvent(packageName, transactionBuilder.getSkuId(),
        transactionBuilder.amount()
            .toString(), BillingAnalytics.PAYMENT_METHOD_REWARDS, transactionBuilder.getType());
  }

  private void sendPaymentErrorEvent(RewardsManager.RewardPayment transaction) {
    if (transaction.getStatus() == RewardsManager.RewardPayment.Status.ERROR
        || transaction.getStatus() == RewardsManager.RewardPayment.Status.NO_NETWORK
        || transaction.getStatus() == RewardsManager.RewardPayment.Status.FORBIDDEN) {
      analytics.sendPaymentErrorEvent(packageName, transactionBuilder.getSkuId(),
          transactionBuilder.amount()
              .toString(), BillingAnalytics.PAYMENT_METHOD_REWARDS, transactionBuilder.getType(),
          transaction.getStatus()
              .toString());
    }
  }

  private void handleSupportClicks() {
    disposables.add(Observable.merge(view.getSupportIconClick(), view.getSupportLogoClick())
        .throttleFirst(50, TimeUnit.MILLISECONDS)
        .flatMapCompletable(__ -> appcoinsRewardsBuyInteract.showSupport(gamificationLevel))
        .subscribe());
  }

  @StringRes private int mapError(RewardsManager.RewardPayment.Status status) {
    if (status == FORBIDDEN) {
      return R.string.purchase_wallet_error_contact_us;
    } else {
      return R.string.activity_iab_error_message;
    }
  }
}

File: app/src/main/java/com/asfoundation/wallet/ui/iab/AppcoinsRewardsBuyPresenter.kt
package com.asfoundation.wallet.ui.iab

import com.appcoins.wallet.appcoins.rewards.Transaction
import com.appcoins.wallet.core.network.microservices.model.BillingSupportedType
import com.appcoins.wallet.core.utils.jvm_common.Logger
import com.asf.wallet.R
import com.asfoundation.wallet.billing.analytics.BillingAnalytics
import com.asfoundation.wallet.entity.TransactionBuilder
import com.appcoins.wallet.core.utils.android_common.CurrencyFormatUtils
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import java.math.BigDecimal
import java.util.concurrent.TimeUnit

class AppcoinsRewardsBuyPresenter(
  private val view: AppcoinsRewardsBuyView,
  private val rewardsManager: RewardsManager,
  private val viewScheduler: Scheduler,
  private val networkScheduler: Scheduler,
  private val disposables: CompositeDisposable,
  private val packageName: String,
  private val isBds: Boolean,
  private val isPreSelected: Boolean,
  private val analytics: BillingAnalytics,
  private val paymentAnalytics: PaymentMethodsAnalytics,
  private val transactionBuilder: TransactionBuilder,
  private val formatter: CurrencyFormatUtils,
  private val gamificationLevel: Int,
  private val appcoinsRewardsBuyInteract: AppcoinsRewardsBuyInteract,
  private val logger: Logger
) {

  companion object {
    private val TAG = AppcoinsRewardsBuyPresenter::class.java.name
  }

  fun present() {
    view.lockRotation()
    handleBuyClick()
    handleOkErrorClick()
    handleSupportClicks()
  }

  private fun handleOkErrorClick() {
    disposables.add(view.getOkErrorClick()
      .doOnNext {
        view.showPaymentMethods()
      }
      .subscribe({}, {
        logger.log(TAG, "Ok error click", it)
        view.errorClose()
      })
    )
  }

  private fun handleBuyClick() {
    disposables.add(rewardsManager
      .pay(
        transactionBuilder.skuId,
        transactionBuilder.amount(),
        transactionBuilder.toAddress(),
        packageName,
        getOrigin(isBds, transactionBuilder),
        transactionBuilder.type,
        transactionBuilder.payload,
        transactionBuilder.callbackUrl,
        transactionBuilder.orderReference,
        transactionBuilder.referrerUrl,
        transactionBuilder.productToken
      )
      .andThen(
        rewardsManager.getPaymentStatus(
          packageName,
          transactionBuilder.skuId,
          transactionBuilder.amount()
        )
      )
      .subscribeOn(networkScheduler)
      .flatMapCompletable {
        handlePaymentStatus(it, transactionBuilder.skuId, transactionBuilder.amount())
      }
      .observeOn(viewScheduler)
      .doOnSubscribe {
        paymentAnalytics.startTimingForPurchaseEvent()
        view.showLoading()
      }
      .doOnError {
        logger.log(TAG, it)
        view.showError(null)
      }
      .subscribe({}, {})
    )
  }

  private fun getOrigin(isBds: Boolean, transaction: TransactionBuilder): String? =
    if (transaction.origin == null) {
      if (isBds) "BDS" else null
    } else {
      transaction.origin
    }

  private fun handlePaymentStatus(
    transaction: RewardPayment,
    sku: String?,
    amount: BigDecimal
  ): Completable {
    sendPaymentErrorEvent(transaction)
    return when (transaction.status) {
      Status.PROCESSING -> Completable.fromAction { view.showLoading() }.subscribeOn(viewScheduler)
      Status.COMPLETED -> {
        if (isBds && isManagedPaymentType(transactionBuilder.type)) {
          val billingType = BillingSupportedType.valueOfProductType(transactionBuilder.type)
          rewardsManager.getPaymentCompleted(packageName, sku, transaction.purchaseUid, billingType)
            .flatMapCompletable { purchase ->
              Completable.fromAction { view.showTransactionCompleted() }
                .subscribeOn(viewScheduler)
                .andThen(Completable.timer(view.getAnimationDuration(), TimeUnit.MILLISECONDS, viewScheduler))
                .andThen(Completable.fromAction { appcoinsRewardsBuyInteract.removeAsyncLocalPayment() })
                .andThen(Completable.fromAction {
                  view.finish(purchase, transaction.orderReference)
                })
            }
            .observeOn(viewScheduler)
            .onErrorResumeNext {
              Completable.fromAction {
                logger.log(TAG, "Error after completing the transaction", it)
                view.showError(null)
                view.hideLoading()
              }
            }
        } else {
          rewardsManager.getTransaction(packageName, sku, amount).firstOrError()
            .map(Transaction::txId).flatMapCompletable { transactionId ->
              Completable.fromAction { view.showTransactionCompleted() }
                .subscribeOn(viewScheduler)
                .andThen(Completable.timer(view.getAnimationDuration(), TimeUnit.MILLISECONDS, viewScheduler))
                .andThen(Completable.fromAction { view.finish(transactionId) })
            }
        }
      }
      Status.ERROR -> Completable.fromAction {
        logger.log(TAG, "Credits error: ${transaction.errorMessage}")
        view.showError(null)
      }.subscribeOn(viewScheduler)
      Status.FORBIDDEN -> Completable.fromAction {
        logger.log(TAG, "Forbidden")
        handleFraudFlow()
      }
      Status.SUB_ALREADY_OWNED -> Completable.fromAction {
        logger.log(TAG, "Sub already owned")
        view.showError(R.string.subscriptions_error_already_subscribed)
      }.subscribeOn(viewScheduler)
      Status.NO_NETWORK -> Completable.fromAction {
        logger.log(TAG, Exception("PaymentStatus no network"))
        view.showNoNetworkError()
        view.hideLoading()
      }.subscribeOn(viewScheduler)
    }
  }

  private fun handleFraudFlow() {
    disposables.add(
      appcoinsRewardsBuyInteract.isWalletBlocked()
        .subscribeOn(networkScheduler)
        .observeOn(networkScheduler)
        .flatMap { blocked ->
          if (blocked) {
            appcoinsRewardsBuyInteract
              .isWalletVerified()
              .observeOn(viewScheduler)
              .doOnSuccess {
                if (it) {
                  logger.log(TAG, Exception("FraudFlow blocked"))
                  view.showError(R.string.purchase_error_wallet_block_code_403)
                } else {
                  view.showVerification()
                }
              }
          } else {
            Single.just(true)
              .observeOn(viewScheduler)
              .doOnSuccess {
                logger.log(TAG, Exception("FraudFlow not blocked"))
                view.showError(R.string.purchase_error_wallet_block_code_403)
              }
          }
        }
        .observeOn(viewScheduler)
        .subscribe({}, {
          logger.log(TAG, it)
          view.showError(R.string.purchase_error_wallet_block_code_403)
        })
    )
  }

  fun stop() = disposables.clear()

  fun sendPaymentEvent() {
    analytics.sendPaymentEvent(
      packageName,
      transactionBuilder.skuId,
      transactionBuilder.amount().toString(),
      BillingAnalytics.PAYMENT_METHOD_REWARDS,
      transactionBuilder.type
    )
  }

  fun sendRevenueEvent() {
    analytics.sendRevenueEvent(
      formatter.scaleFiat(
        appcoinsRewardsBuyInteract.convertToFiat(
          transactionBuilder.amount().toDouble(),
          BillingAnalytics.EVENT_REVENUE_CURRENCY
        ).blockingGet()
          .amount
      ).toString()
    )
  }

  fun sendPaymentSuccessEvent(txId: String) {
    paymentAnalytics.stopTimingForPurchaseEvent(
      PaymentMethodsAnalytics.PAYMENT_METHOD_APPC,
      true,
      isPreSelected
    )
    analytics.sendPaymentSuccessEvent(
      packageName = packageName,
      skuDetails = transactionBuilder.skuId,
      value = transactionBuilder.amount().toString(),
      purchaseDetails = BillingAnalytics.PAYMENT_METHOD_REWARDS,
      transactionType = transactionBuilder.type,
      txId = txId,
      valueUsd = transactionBuilder.amountUsd.toString()
    )
  }

  private fun sendPaymentErrorEvent(transaction: RewardPayment) {
    val status = transaction.status
    if (isErrorStatus(status)) {
      paymentAnalytics.stopTimingForPurchaseEvent(
        PaymentMethodsAnalytics.PAYMENT_METHOD_APPC,
        true,
        isPreSelected
      )
      if (transaction.errorCode == null && transaction.errorMessage == null) {
        analytics.sendPaymentErrorEvent(
          packageName,
          transactionBuilder.skuId,
          transactionBuilder.amount().toString(),
          BillingAnalytics.PAYMENT_METHOD_REWARDS,
          transactionBuilder.type,
          status.toString()
        )
      } else {
        analytics.sendPaymentErrorWithDetailsEvent(
          packageName,
          transactionBuilder.skuId,
          transactionBuilder.amount().toString(),
          BillingAnalytics.PAYMENT_METHOD_REWARDS,
          transactionBuilder.type,
          transaction.errorCode.toString(),
          transaction.errorMessage.toString()
        )
      }
    }
  }

  private fun isErrorStatus(status: Status): Boolean =
    status === Status.ERROR || status === Status.NO_NETWORK || status === Status.FORBIDDEN || status === Status.SUB_ALREADY_OWNED

  private fun handleSupportClicks() {
    disposables.add(
      Observable.merge(view.getSupportIconClick(), view.getSupportLogoClick())
        .throttleFirst(50, TimeUnit.MILLISECONDS)
        .observeOn(viewScheduler)
        .flatMapCompletable { appcoinsRewardsBuyInteract.showSupport(gamificationLevel) }
        .subscribe({}, { it.printStackTrace() })
    )
  }

  private fun isManagedPaymentType(type: String): Boolean =
    type == BillingSupportedType.INAPP.name || type == BillingSupportedType.INAPP_SUBSCRIPTION.name
}